Lab_1-2-3/lab-1-2-3-sql-window-function.md

```diff
@@ -1306,222 +1367,207 @@ Zbiór wynikowy powinien zawierać:
-    ph.date;
+    d.*,
+    sum(d2.dayValue) over(partition by d.productid, strftime('%Y %m', d.date)
+    order by strftime('%d', d.date) ) as accumulated
+FROM Data as d
+JOIN Data as d2 on d.productid = d2.productid and
+                   strftime('%Y %m', d.date) = strftime('%Y %m', d2.date) and
+                   strftime('%d', d.date) >= strftime('%d', d2.date)
+ORDER BY d.date
 ```
 
-> Dla żadnego SZBD zapytanie nie skończyło się w rozsądnym czasie, więc byliśmy zmuszeni do ogranizcenia ilości danych do 1000 wierszy. Brak wyników spowodowany jest użyciem bardzo dużego subquery.
-
-**Wyniki**
+**Czasy wykonania**
 
-| MS SQL                     | Postgres                    | SQLite                    |
-| -------------------------- | --------------------------- | ------------------------- |
-| ![](./img/ex14/mysql1.png) | ![](img/ex14/postgres1.png) | ![](img/ex14/sqlite1.png) |
+| Z funkcją okna lub bez | MS SQL                         | Postgres                          | SQLite                          |
+| ---------------------- | ------------------------------ | --------------------------------- | ------------------------------- |
+| Z funkcją okna         | ![](./img/ex14/mssql1time.png) | ![](./img/ex14/postgres1time.png) | ![](./img/ex14/sqlite1time.png) |
+| Bez funkcji okna       | 10 min + (przerwane)           | 10 min + (przerwane)              | 10 min + (przerwane)            |
 
 **Plany wykonania**
 
-| MS SQL                         | Postgres                        | SQLite                        |
-| ------------------------------ | ------------------------------- | ----------------------------- |
-| ![](./img/ex14/mysql1plan.png) | ![](img/ex14/postgres1plan.png) | ![](img/ex14/sqlite1plan.png) |
+| Z funkcją okna lub bez | MS SQL                         | Postgres                        | SQLite                        |
+| ---------------------- | ------------------------------ | ------------------------------- | ----------------------------- |
+| Z funkcją okna         | ![](./img/ex14/mssql1plan.png) | ![](img/ex14/postgres1plan.png) | ![](img/ex14/sqlite1plan.png) |
+| Bez funkcji okna       | ![](./img/ex14/mssql2plan.png) | ![](img/ex14/postgres2plan.png) | ![](img/ex14/sqlite2plan.png) |
 
-> Jak widzimy plan wykonania dla MS SQL jest znacznie bardziej rozbudowany niż dla SQLite czy Postgres. Można tez zauważyć różnicę w kosztach zapytań. Podzapytania te są bardzo skomplikowane, przez co ich koszt jest bardzo duży. Operacja filtrowania nie może być uproszczona i to powoduje wolne działanie całego zapytania.
+> Dla zapytania bez funkcji okna PostgreSQL wydaje się lepszy od SQL Servera pod względem równoległego wykonywania operacji w planie wykonania, co przekłada się na rzeczywisty czas wykonania. SQLite w tym przypadku nie radził sobie dobrze z zapytaniem o takim stopniu skomplikowania i takiej liczbie wierszy. Wynikowy czas wykonania zapytania z funkcją okna może być akceptowalny, ale już bez niej nie.
```

Comment:
można opisać, że równoległa operacja to "Paralellism (Gather Streams)" a bez okna mamy skany indeksów, osobne obliczanie skalarów i kosztowne agregacje

Reply:
Dodane